### ISSUE TITLE
Inconsistencies between hand-calculated gradients and machine-calculated

### ISSUE THREAD

**Issue by paukey** [clab/dynet#1663]
I defined a mlp netword and loss function squared_distance。

the loss value computed by Dynet is identical with hand-calculated。

but，after running backward function，the gradient is not identical with hand-calculated。

The Dynet get the gradient by standrd BP algorithm？？

BP： https://en.wikipedia.org/wiki/Backpropagation
